ARROW-10191: [Rust] [Parquet] Add roundtrip Arrow -> Parquet tests for all supported Arrow DataTypes

carols10cents opened the pull request:
Note that this PR goes to the rust-parquet-arrow-writer branch, not master.

Inspired by tests in cpp/src/parquet/arrow/arrow_reader_writer_test.cc

These perform round-trip Arrow -> Parquet -> Arrow of a single RecordBatch with a single column of values of each the supported data types and some of the unsupported ones.

Tests that currently fail are either marked with `#[should_panic]` (if the
reason they fail is because of a panic) or `#[ignore]` (if the reason
they fail is because the values don't match).

I am comparing the RecordBatch's column's data before and after the round trip directly; I'm not sure that this is appropriate or not because for some data types, the `null_bitmap` isn't matching and I'm not sure if it's supposed to or not.

So I would love advice on that front, and I would love to know if these tests are useful or not!

carols10cents commented:
cc @nevi-me

alamb commented:
> I am comparing the RecordBatch's column's data before and after the round trip directly; I'm not sure that this is appropriate or not because for some data types, the null_bitmap isn't matching and I'm not sure if it's supposed to or not.

I think the null bitmap should definitely match after the round trip (as it signals which rows are null, which should be preserved).

I am not sure the actual data arrays have to be the same -- what needs to be the same is the data rows where the null_bit is 0 (aka we should probably be checking only for non-null rows)

alamb reviewed (approved):
FWIW I think the basic idea and structure of these tests  is ✔️  -- with the possible exception of directly comparing the data values regardless of the null bit vector, I think the test are good.

carols10cents commented on rust/parquet/src/arrow/arrow_writer.rs:
```diff
@@ -688,4 +688,413 @@ mod tests {
+    }
+
+    #[test]
+    #[ignore] // Large Binary support isn't correct yet
```
This test currently fails because both `LargeBinary` and `Binary` Arrow types are converted to Parquet `BYTE_ARRAY` types. I think this is what's intended, and I can't find any C++ tests that specifically exercise Arrow LargeBinary data, so I'm not sure if it's just supposed to roundtrip and lose the "large" information or if it's supposed to panic as not supported either on writing or reading.

```
---- arrow::arrow_writer::tests::large_binary_single_column stdout ----
thread 'arrow::arrow_writer::tests::large_binary_single_column' panicked at 'assertion failed: `(left == right)`
  left: `Schema { fields: [Field { name: "col", data_type: LargeBinary, nullable: false, dict_id: 0, dict_is_ordered: false }], metadata: {} }`,
 right: `Schema { fields: [Field { name: "col", data_type: Binary, nullable: false, dict_id: 0, dict_is_ordered: false }], metadata: {} }`', parquet/src/arrow/arrow_writer.rs:712:9
```

emkornfield replied:
I'm currently working on adding some of this support in C++.  Based on parquet metadata there is not a great way to make distinguish these types.  The way C++ handles it is on serialization the Arrow schema for the table is added as metadata (its base64 encoded).  When reading back [we deserialize](https://github.com/apache/arrow/blob/477c1021ac013f22389baf9154fb9ad0cf814bec/cpp/src/parquet/arrow/schema.cc#L653) it and use the types there to make an additional determination.  Currently we only use it for [extension types](https://github.com/apache/arrow/blob/477c1021ac013f22389baf9154fb9ad0cf814bec/cpp/src/parquet/arrow/schema.cc#L692)

nevi-me replied:
In this case, the reader is supposed to get the correct type from the serialized schema, and then do the conversion to a large array

carols10cents commented on rust/parquet/src/arrow/arrow_writer.rs:
```diff
@@ -688,4 +688,413 @@ mod tests {
+    }
+
+    #[test]
+    #[ignore] // Large String support isn't correct yet - null_bitmap and buffers don't match
```
This is actually the same as LargeArray-- the Large part of LargeUtf8 isn't coming back, and I'm not sure the intended way to handle that.

```
---- arrow::arrow_writer::tests::large_string_single_column stdout ----
thread 'arrow::arrow_writer::tests::large_string_single_column' panicked at 'assertion failed: `(left == right)`
  left: `Schema { fields: [Field { name: "col", data_type: LargeUtf8, nullable: false, dict_id: 0, dict_is_ordered: false }], metadata: {} }`,
 right: `Schema { fields: [Field { name: "col", data_type: Utf8, nullable: false, dict_id: 0, dict_is_ordered: false }], metadata: {} }`', parquet/src/arrow/arrow_writer.rs:712:9
```

nevi-me replied:
May you please have a look at the function `parquet_to_arrow_schema_by_columns` and https://issues.apache.org/jira/browse/ARROW-10168. I suspect this might be why the arrow metadata gets lost during the roundtrip.

carols10cents reviewed (commented):
Adding some notes about why some tests are ignored/failing

emkornfield commented:
> I am not sure the actual data arrays have to be the same -- what needs to be the same is the data rows where the null_bit is 0 (aka we should probably be checking only for non-null rows)

Small nit I'm prettry sure  null_bit = 1 is actually where values should be compared.  (thinking of it as a validity bitmap instead of a null bitmap helps).  And yes I agree they should be compared separately, as there is generally no requirement on the values of null fields (generally, you'll want to initialize to some known values to avoid security issues).

nevi-me commented on rust/parquet/src/arrow/arrow_writer.rs:
```diff
@@ -688,4 +688,413 @@ mod tests {
+    fn struct_single_column() {
+        let a_values = Int32Array::from(vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
+        let struct_field_a = Field::new("f", DataType::Int32, false);
+        let s = StructArray::from(vec![(struct_field_a, Arc::new(a_values) as ArrayRef)]);
```
The cause might be this here. See https://issues.apache.org/jira/browse/ARROW-5408.

nevi-me reviewed (commented):
Hi @carols10cents, I've added some notes on why some of the tests are failing. I opened https://github.com/integer32llc/arrow/pull/1. I ran out of time, I could have looked at other ignored tests, but I suppose running out of time is why I need the help :)

nevi-me commented:
> And yes I agree they should be compared separately, as there is generally no requirement on the values of null fields (generally, you'll want to initialize to some known values to avoid security issues).

There was a time ~2 years ago where null slots would contain garbage/unknown values, but I haven't seen that in very long.
With nested arrays, we'll also want to convert the data directly as a struct with null slots containing a primitive that's non-nullable, would result in a nullable primitive after roundtrip.

@emkornfield in the above example, does the c++ Arrow implementation allow creating such a struct? Asked differently; is it some validation that we should be adding at an Arrow level in Rust, that if you create a nullable struct, you can't have its children being non-nullable?

nevi-me commented:
I rebased the parquet branch against the main one, but now I seem to have picked up additional commits on this PR. They should be fixed by a rebase. @carols10cents may you please grant me write permission on Integer32's fork, so I can rebase there.

carols10cents commented:
@nevi-me you should have an invite!

carols10cents commented:
@nevi-me I went ahead and merged in [your PR](https://github.com/integer32llc/arrow/pull/1), rebased this PR's branch on rust-parquet-arrow-writer, added one commit to explicitly not support Float16 in the `match` in `write_leaves`, and force pushed. Do you think this can be merged or do more of the ignored tests need to be passing? I'm going to take a look at reading the schema metadata next, in any case.

nevi-me commented:
> @nevi-me I went ahead and merged in [your PR](https://github.com/integer32llc/arrow/pull/1), rebased this PR's branch on rust-parquet-arrow-writer, added one commit to explicitly not support Float16 in the `match` in `write_leaves`, and force pushed. Do you think this can be merged or do more of the ignored tests need to be passing? I'm going to take a look at reading the schema metadata next, in any case.

I'll merge this in my morning (GMT+2). I'm happy with the tests, they also help with unitising further work so others can help us out. I'll open JIRAs for the cases that still fail

nevi-me commented:
@carols10cents may you please create a JIRA account at https://issues.apache.org, so that we can assign the tasks that you've worked on to you. Thanks :)

nevi-me commented:
Hi @andygrove @jorgecarleitao may you please kindly merge this for me, the merge tool's failing to push to GH on my end :(

carols10cents commented:
@nevi-me Done! https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carols10cents

andygrove commented:
@nevi-me @carols10cents I ran the merge tool and the changes appear to have been committed to the branch but this issue did not get closed, so I am closing it manually.

nevi-me commented:
> @nevi-me @carols10cents I ran the merge tool and the changes appear to have been committed to the branch but this PR did not get closed, so I am closing it manually.

It's been the case since I started merging on the branch. I've been meaning to open a JIRA for this, but keep forgetting. I've opened https://issues.apache.org/jira/browse/ARROW-10198
